data_structures/stack_using_array.cpp

```diff
@@ -1,62 +1,178 @@
-#include <iostream>
+/**
+ * @file
+ * @brief Implementation of a stack data structure
+ * @details
+ * This implementation provides functionalities to push, pop, and view elements
+ * of the stack. It also includes a self-test method to ensure proper
+ * functionality.
+ */
 
-int *stack;
-int stack_idx = 0, stack_size;
+#include <cassert>    /// For assert
+#include <iostream>   /// For IO operations
+#include <stdexcept>  /// For std::out_of_range
 
-void push(int x) {
-    if (stack_idx == stack_size) {
-        std::cout << "\nOverflow";
-    } else {
-        stack[stack_idx++] = x;
+/*
+ * @namespace
+ * @brief Data structures
+ */
+namespace data_structures {
+/**
+ * @brief Class representation of a stack
+ * @tparam T The type of the elements in the stack
+ */
+template <typename T>
+class Stack {
+ private:
+    T* stack;        ///< Pointer to the stack array
+    int stackSize;   ///< Maximum size of the stack
+    int stackIndex;  ///< Index pointing to the top element of the stack
+
+ public:
+    /**
+     * @brief Constructs a new Stack object
+     *
+     * @param size Maximum size of the stack
+     */
+    Stack(int size) : stackSize(size), stackIndex(0) { stack = new T[size]; }
```

Comment:
```suggestion
    Stack(int size) : stackSize(size), stackIndex(-1) { stack = new T[size]; }
```
TOS should be -1 at the start